### PR TITLE
*: implement subquery expressions

### LINKED ISSUE #59: Support sqlparser.Subquery statement

### PR DESCRIPTION
**Pull request by erizocosmico**
Closes #59

### CODE REVIEW

File: sql/expression/subquery_test.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+		plan.NewResolvedTable(table),
+	))
+
+	values, err := subquery.EvalMultiple(ctx)
+	require.NoError(err)
+	require.Equal(values, []interface{}{"one", "two", "three"})
+
+	_, err = subquery.Eval(ctx, nil)
+	require.Error(err)
+}
```
[the remark this answers is not in the document]
Comment (erizocosmico) [reply sql/expression/subquery_test.go]: This is actually not needed, it's tested in another place

File: sql/expression/subquery.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+func (s *Subquery) Resolved() bool {
+	return s.Query.Resolved()
+}
+
+// Type implements the Expression interface.
+func (s *Subquery) Type() sql.Type {
+	return s.Query.Schema()[0].Type
+}
+
+// WithChildren implements the Expression interface.
```
[the remark this answers is not in the document]
Comment (erizocosmico) [reply sql/expression/subquery.go]: This is just to implement the Expression interface, yes

File: sql/analyzer/validation_rules.go [hunk 3]
```diff
@@ -322,6 +330,24 @@ func validateExplodeUsage(ctx *sql.Context, a *Analyzer, n sql.Node) (sql.Node,
 	return n, nil
 }
 
+func validateSubqueryColumns(ctx *sql.Context, a *Analyzer, n sql.Node) (sql.Node, error) {
+	valid := true
+	plan.InspectExpressions(n, func(e sql.Expression) bool {
+		s, ok := e.(*expression.Subquery)
+		if ok && len(s.Query.Schema()) != 1 {
+			valid = false
```
[the remark this answers is not in the document]
Comment (erizocosmico) [reply sql/analyzer/validation_rules.go]: correct

File: sql/analyzer/assign_indexes.go [hunk 4]
```diff
@@ -759,8 +759,19 @@ func containsColumns(e sql.Expression) bool {
 	return result
 }
 
+func containsSubquery(e sql.Expression) bool {
+	var result bool
+	expression.Inspect(e, func(e sql.Expression) bool {
+		if _, ok := e.(*expression.Subquery); ok {
+			result = true
```
[the remark this answers is not in the document]
Comment (erizocosmico) [reply sql/analyzer/assign_indexes.go]: correctr

File: sql/parse/parse.go [hunk 5]
```diff
@@ -135,7 +132,10 @@ func convert(ctx *sql.Context, stmt sqlparser.Statement, query string) (sql.Node
 	default:
 		return nil, ErrUnsupportedSyntax.New(n)
 	case *sqlparser.Show:
-		return convertShow(n, query)
+		if query == "" {
+			return nil, ErrUnsupportedFeature.New("SHOW in subquery")
+		}
```
[the remark this answers is not in the document]
Comment (erizocosmico) [reply sql/parse/parse.go]: We don't have the subquery query as text, which is needed to do some manual parsing for SHOW. I've added a comment explaining that.

File: sql/expression/subquery.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (erizocosmico) [reply sql/expression/subquery.go]: I've added a check that i missed that will return an error when the children length is not zero.

**Comment by erizocosmico** [comment]
Done

File: sql/analyzer/analyzer.go [hunk 6]
```diff
@@ -1,4 +1,4 @@
-package analyzer // import "github.com/src-d/go-mysql-server/sql/analyzer"
```
Comment (ajnavarro) [inline sql/analyzer/analyzer.go]: these comments are not needed anymore?

Comment (erizocosmico) [reply sql/analyzer/analyzer.go]: nope

**Comment by erizocosmico** [comment]
Changed supported.md

**Comment by ajnavarro** [comment]
@erizocosmico could you rebase to fix CI? thanks!

**Pull request merged**